File: study_planner.py
# study_planner.py
import json
from dataclasses import dataclass, asdict
from enum import Enum
from datetime import datetime, timedelta
from typing import List, Optional
from pathlib import Path

class Priority(Enum):
    LOW = "Низкий"
    MEDIUM = "Средний"
    HIGH = "Высокий"

class SessionStatus(Enum):
    PLANNED = "Запланирована"
    COMPLETED = "Выполнена"
    MISSED = "Пропущена"

@dataclass
class StudySession:
    id: int
    subject: str
    topic: str
    datetime: str  # "YYYY-MM-DD HH:MM"
    duration: int   # minutes
    priority: Priority
    status: SessionStatus

class StudyPlanner:
    def __init__(self):
        self.sessions: List[StudySession] = []
        self.next_id = 1

    def add_session(self, subject: str, topic: str, dt: str, duration: int, priority: Priority) -> StudySession:
        # Проверка формата даты
        try:
            datetime.strptime(dt, "%Y-%m-%d %H:%M")
        except ValueError:
            raise ValueError("Неверный формат даты/времени, используйте ГГГГ-ММ-ДД ЧЧ:ММ")
        if duration <= 0:
            raise ValueError("Длительность должна быть положительной")
        session = StudySession(
            id=self.next_id,
            subject=subject,
            topic=topic,
            datetime=dt,
            duration=duration,
            priority=priority,
            status=SessionStatus.PLANNED
        )
        self.sessions.append(session)
        self.next_id += 1
        return session

    def find_session(self, session_id: int) -> Optional[StudySession]:
        return next((s for s in self.sessions if s.id == session_id), None)

    def update_session(self, session_id: int, **kwargs) -> bool:
        session = self.find_session(session_id)
        if not session:
            return False
        for key, value in kwargs.items():
            if hasattr(session, key) and value is not None:
                setattr(session, key, value)
        return True

    def delete_session(self, session_id: int) -> bool:
        session = self.find_session(session_id)
        if session:
            self.sessions.remove(session)
            return True
        return False

    def set_status(self, session_id: int, status: SessionStatus) -> bool:
        session = self.find_session(session_id)
        if not session:
            return False
        session.status = status
        return True

    def filter_by_subject(self, subject: str) -> List[StudySession]:
        return [s for s in self.sessions if s.subject.lower() == subject.lower()]

    def filter_by_status(self, status: SessionStatus) -> List[StudySession]:
        return [s for s in self.sessions if s.status == status]

    def filter_by_date(self, date_str: str) -> List[StudySession]:
        return [s for s in self.sessions if s.datetime.startswith(date_str)]

    def get_stats(self) -> dict:
        total = len(self.sessions)
        completed = len(self.filter_by_status(SessionStatus.COMPLETED))
        planned = len(self.filter_by_status(SessionStatus.PLANNED))
        missed = len(self.filter_by_status(SessionStatus.MISSED))
        total_time = sum(s.duration for s in self.sessions)
        completed_time = sum(s.duration for s in self.sessions if s.status == SessionStatus.COMPLETED)
        # Статистика по предметам
        subjects = {}
        for s in self.sessions:
            subjects[s.subject] = subjects.get(s.subject, 0) + s.duration
        avg_duration = total_time / total if total > 0 else 0
        return {
            "total": total,
            "completed": completed,
            "planned": planned,
            "missed": missed,
            "total_time": total_time,
            "completed_time": completed_time,
            "avg_duration": avg_duration,
            "by_subject": subjects
        }

    def save_to_file(self, filename: str = "sessions_data.json") -> None:
        data = {"sessions": [asdict(s) for s in self.sessions]}
        with open(filename, "w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False, indent=2)

    def load_from_file(self, filename: str = "sessions_data.json") -> None:
        path = Path(filename)
        if not path.exists():
            return
        with open(filename, "r", encoding="utf-8") as f:
            data = json.load(f)
            self.sessions.clear()
            for item in data.get("sessions", []):
                session = StudySession(
                    id=item["id"],
                    subject=item["subject"],
                    topic=item["topic"],
                    datetime=item["datetime"],
                    duration=item["duration"],
                    priority=Priority(item["priority"]),
                    status=SessionStatus(item["status"])
                )
                self.sessions.append(session)
                if session.id >= self.next_id:
                    self.next_id = session.id + 1

def print_session(session: StudySession) -> None:
    status_emoji = {"Запланирована": "📅", "Выполнена": "✅", "Пропущена": "❌"}
    print(f"{status_emoji[session.status.value]} #{session.id} - {session.subject} ({session.topic})")
    print(f"   Дата/время: {session.datetime}, Длительность: {session.duration} мин")
    print(f"   Приоритет: {session.priority.value}, Статус: {session.status.value}")

def main():
    planner = StudyPlanner()
    planner.load_from_file()

    while True:
        print("\n===== ПЛАНИРОВЩИК УЧЕБНЫХ СЕССИЙ (Python) =====")
        print("1. Добавить сессию")
        print("2. Показать все сессии")
        print("3. Показать сессии по предмету")
        print("4. Отметить сессию как выполненную")
        print("5. Редактировать сессию")
        print("6. Удалить сессию")
        print("7. Показать статистику")
        print("8. Сохранить в файл")
        print("9. Загрузить из файла")
        print("0. Выход")
        choice = input("Выберите действие: ").strip()

        if choice == "0":
            break
        elif choice == "1":
            subject = input("Предмет: ").strip()
            if not subject:
                print("Предмет не может быть пустым.")
                continue
            topic = input("Тема: ").strip()
            dt = input("Дата и время (ГГГГ-ММ-ДД ЧЧ:ММ): ").strip()
            try:
                dur = int(input("Длительность (мин): ").strip())
            except ValueError:
                print("Введите число.")
                continue
            print("Приоритет: 1-Низкий, 2-Средний, 3-Высокий")
            prio_input = input("> ").strip()
            if prio_input == "1":
                priority = Priority.LOW
            elif prio_input == "3":
                priority = Priority.HIGH
            else:
                priority = Priority.MEDIUM
            try:
                session = planner.add_session(subject, topic, dt, dur, priority)
                print(f"Сессия добавлена с ID {session.id}")
            except Exception as e:
                print("Ошибка:", e)
        elif choice == "2":
            if not planner.sessions:
                print("Нет сессий.")
            else:
                for s in planner.sessions:
                    print_session(s)
        elif choice == "3":
            subject = input("Введите предмет: ").strip()
            sessions = planner.filter_by_subject(subject)
            if not sessions:
                print("Сессий по этому предмету нет.")
            else:
                for s in sessions:
                    print_session(s)
        elif choice == "4":
            try:
                sid = int(input("Введите ID сессии: ").strip())
            except ValueError:
                print("Некорректный ID.")
                continue
            if planner.set_status(sid, SessionStatus.COMPLETED):
                print("Сессия отмечена как выполненная.")
            else:
                print("Сессия не найдена.")
        elif choice == "5":
            try:
                sid = int(input("Введите ID сессии для редактирования: ").strip())
            except ValueError:
                print("Некорректный ID.")
                continue
            session = planner.find_session(sid)
            if not session:
                print("Сессия не найдена.")
                continue
            print("Оставьте поле пустым, чтобы не менять.")
            new_subj = input(f"Предмет ({session.subject}): ").strip()
            new_topic = input(f"Тема ({session.topic}): ").strip()
            new_dt = input(f"Дата/время ({session.datetime}): ").strip()
            new_dur = input(f"Длительность ({session.duration}): ").strip()
            new_prio = input(f"Приоритет (1-Низкий,2-Средний,3-Высокий) сейчас: {session.priority.value}: ").strip()
            updates = {}
            if new_subj: updates["subject"] = new_subj
            if new_topic: updates["topic"] = new_topic
            if new_dt: updates["datetime"] = new_dt
            if new_dur:
                try:
                    updates["duration"] = int(new_dur)
                except ValueError:
                    print("Длительность должна быть числом, пропускаем.")
            if new_prio:
                if new_prio == "1": updates["priority"] = Priority.LOW
                elif new_prio == "3": updates["priority"] = Priority.HIGH
                else: updates["priority"] = Priority.MEDIUM
            if planner.update_session(sid, **updates):
                print("Сессия обновлена.")
            else:
                print("Ошибка обновления.")
        elif choice == "6":
            try:
                sid = int(input("Введите ID сессии для удаления: ").strip())
            except ValueError:
                print("Некорректный ID.")
                continue
            if planner.delete_session(sid):
                print("Сессия удалена.")
            else:
                print("Сессия не найдена.")
        elif choice == "7":
            stats = planner.get_stats()
            print("\n=== СТАТИСТИКА ===")
            print(f"Всего сессий: {stats['total']}")
            print(f"Выполнено: {stats['completed']}")
            print(f"Запланировано: {stats['planned']}")
            print(f"Пропущено: {stats['missed']}")
            print(f"Общее время: {stats['total_time']} мин")
            print(f"Время выполненных: {stats['completed_time']} мин")
            print(f"Средняя длительность: {stats['avg_duration']:.1f} мин")
            print("По предметам:")
            for subj, minutes in stats['by_subject'].items():
                print(f"  {subj}: {minutes} мин")
        elif choice == "8":
            planner.save_to_file()
            print("Сохранено.")
        elif choice == "9":
            planner.load_from_file()
            print("Загружено.")
        else:
            print("Неизвестная команда.")

if __name__ == "__main__":
    main()

File: study_planner.rb
# study_planner.rb
require 'json'
require 'date'

class Priority
  LOW = "Низкий"
  MEDIUM = "Средний"
  HIGH = "Высокий"
end

class SessionStatus
  PLANNED = "Запланирована"
  COMPLETED = "Выполнена"
  MISSED = "Пропущена"
end

class StudySession
  attr_accessor :id, :subject, :topic, :datetime, :duration, :priority, :status

  def initialize(id, subject, topic, datetime, duration, priority = Priority::MEDIUM, status = SessionStatus::PLANNED)
    @id = id
    @subject = subject
    @topic = topic
    @datetime = datetime
    @duration = duration
    @priority = priority
    @status = status
  end

  def to_h
    {
      id: @id,
      subject: @subject,
      topic: @topic,
      datetime: @datetime,
      duration: @duration,
      priority: @priority,
      status: @status
    }
  end

  def self.from_h(hash)
    StudySession.new(hash[:id], hash[:subject], hash[:topic], hash[:datetime],
                     hash[:duration], hash[:priority], hash[:status])
  end
end

class StudyPlanner
  attr_reader :sessions

  def initialize
    @sessions = []
    @next_id = 1
  end

  def add_session(subject, topic, datetime, duration, priority = Priority::MEDIUM)
    # Проверка формата даты
    begin
      DateTime.strptime(datetime, "%Y-%m-%d %H:%M")
    rescue ArgumentError
      raise "Неверный формат даты/времени, используйте ГГГГ-ММ-ДД ЧЧ:ММ"
    end
    raise "Длительность должна быть положительной" if duration <= 0
    session = StudySession.new(@next_id, subject, topic, datetime, duration, priority)
    @sessions << session
    @next_id += 1
    session
  end

  def find_session(id)
    @sessions.find { |s| s.id == id }
  end

  def update_session(id, **kwargs)
    session = find_session(id)
    return false unless session
    kwargs.each do |key, value|
      session.send("#{key}=", value) if session.respond_to?("#{key}=")
    end
    true
  end

  def delete_session(id)
    session = find_session(id)
    return false unless session
    @sessions.delete(session)
    true
  end

  def set_status(id, status)
    session = find_session(id)
    return false unless session
    session.status = status
    true
  end

  def filter_by_subject(subject)
    @sessions.select { |s| s.subject.downcase == subject.downcase }
  end

  def filter_by_status(status)
    @sessions.select { |s| s.status == status }
  end

  def filter_by_date(date_str)
    @sessions.select { |s| s.datetime.start_with?(date_str) }
  end

  def stats
    total = @sessions.size
    completed = filter_by_status(SessionStatus::COMPLETED).size
    planned = filter_by_status(SessionStatus::PLANNED).size
    missed = filter_by_status(SessionStatus::MISSED).size
    total_time = @sessions.sum(&:duration)
    completed_time = filter_by_status(SessionStatus::COMPLETED).sum(&:duration)
    avg_duration = total > 0 ? total_time.to_f / total : 0
    by_subject = Hash.new(0)
    @sessions.each { |s| by_subject[s.subject] += s.duration }
    {
      total: total,
      completed: completed,
      planned: planned,
      missed: missed,
      total_time: total_time,
      completed_time: completed_time,
      avg_duration: avg_duration,
      by_subject: by_subject
    }
  end

  def save_to_file(filename = "sessions_data.json")
    data = { sessions: @sessions.map(&:to_h) }
    File.write(filename, JSON.pretty_generate(data))
  end

  def load_from_file(filename = "sessions_data.json")
    return unless File.exist?(filename)
    data = JSON.parse(File.read(filename), symbolize_names: true)
    @sessions.clear
    data[:sessions].each do |item|
      session = StudySession.from_h(item)
      @sessions << session
      @next_id = session.id + 1 if session.id >= @next_id
    end
  rescue JSON::ParserError
    puts "Ошибка чтения файла."
  end
end

def print_session(session)
  emoji = { "Запланирована" => "📅", "Выполнена" => "✅", "Пропущена" => "❌" }
  puts "#{emoji[session.status]} ##{session.id} - #{session.subject} (#{session.topic})"
  puts "   Дата/время: #{session.datetime}, Длительность: #{session.duration} мин"
  puts "   Приоритет: #{session.priority}, Статус: #{session.status}"
end

def main
  planner = StudyPlanner.new
  planner.load_from_file

  loop do
    puts "\n===== ПЛАНИРОВЩИК УЧЕБНЫХ СЕССИЙ (Ruby) ====="
    puts "1. Добавить сессию"
    puts "2. Показать все сессии"
    puts "3. Показать сессии по предмету"
    puts "4. Отметить сессию как выполненную"
    puts "5. Редактировать сессию"
    puts "6. Удалить сессию"
    puts "7. Показать статистику"
    puts "8. Сохранить в файл"
    puts "9. Загрузить из файла"
    puts "0. Выход"
    print "Выберите действие: "
    choice = gets.chomp

    case choice
    when "0"
      break
    when "1"
      print "Предмет: "
      subject = gets.chomp
      next if subject.empty?
      print "Тема: "
      topic = gets.chomp
      print "Дата и время (ГГГГ-ММ-ДД ЧЧ:ММ): "
      datetime = gets.chomp
      print "Длительность (мин): "
      duration = gets.chomp.to_i
      puts "Приоритет: 1-Низкий, 2-Средний, 3-Высокий"
      prio = gets.chomp
      priority = case prio
                 when "1" then Priority::LOW
                 when "3" then Priority::HIGH
                 else Priority::MEDIUM
                 end
      begin
        session = planner.add_session(subject, topic, datetime, duration, priority)
        puts "Сессия добавлена с ID #{session.id}"
      rescue => e
        puts "Ошибка: #{e.message}"
      end
    when "2"
      if planner.sessions.empty?
        puts "Нет сессий."
      else
        planner.sessions.each { |s| print_session(s) }
      end
    when "3"
      print "Введите предмет: "
      subject = gets.chomp
      sessions = planner.filter_by_subject(subject)
      if sessions.empty?
        puts "Сессий по этому предмету нет."
      else
        sessions.each { |s| print_session(s) }
      end
    when "4"
      print "Введите ID сессии: "
      id = gets.chomp.to_i
      if planner.set_status(id, SessionStatus::COMPLETED)
        puts "Сессия отмечена как выполненная."
      else
        puts "Сессия не найдена."
      end
    when "5"
      print "Введите ID сессии для редактирования: "
      id = gets.chomp.to_i
      session = planner.find_session(id)
      unless session
        puts "Сессия не найдена."
        next
      end
      puts "Оставьте поле пустым, чтобы не менять."
      print "Предмет (#{session.subject}): "
      new_subj = gets.chomp
      print "Тема (#{session.topic}): "
      new_topic = gets.chomp
      print "Дата/время (#{session.datetime}): "
      new_dt = gets.chomp
      print "Длительность (#{session.duration}): "
      new_dur = gets.chomp
      print "Приоритет (1-Низкий,2-Средний,3-Высокий) сейчас: #{session.priority}: "
      new_prio = gets.chomp
      updates = {}
      updates[:subject] = new_subj unless new_subj.empty?
      updates[:topic] = new_topic unless new_topic.empty?
      updates[:datetime] = new_dt unless new_dt.empty?
      unless new_dur.empty?
        updates[:duration] = new_dur.to_i
      end
      unless new_prio.empty?
        updates[:priority] = case new_prio
                             when "1" then Priority::LOW
                             when "3" then Priority::HIGH
                             else Priority::MEDIUM
                             end
      end
      if planner.update_session(id, **updates)
        puts "Сессия обновлена."
      else
        puts "Ошибка обновления."
      end
    when "6"
      print "Введите ID сессии для удаления: "
      id = gets.chomp.to_i
      if planner.delete_session(id)
        puts "Сессия удалена."
      else
        puts "Сессия не найдена."
      end
    when "7"
      stats = planner.stats
      puts "\n=== СТАТИСТИКА ==="
      puts "Всего сессий: #{stats[:total]}"
      puts "Выполнено: #{stats[:completed]}"
      puts "Запланировано: #{stats[:planned]}"
      puts "Пропущено: #{stats[:missed]}"
      puts "Общее время: #{stats[:total_time]} мин"
      puts "Время выполненных: #{stats[:completed_time]} мин"
      puts "Средняя длительность: #{'%.1f' % stats[:avg_duration]} мин"
      puts "По предметам:"
      stats[:by_subject].each { |subj, mins| puts "  #{subj}: #{mins} мин" }
    when "8"
      planner.save_to_file
      puts "Сохранено."
    when "9"
      planner.load_from_file
      puts "Загружено."
    else
      puts "Неизвестная команда."
    end
  end
end

main if __FILE__ == $0
